File: python/euler/bin/p0067.py
# project euler: problem 67

from collections.abc import Callable
from functools import reduce
from itertools import pairwise
from operator import add
from typing import TextIO


def parse_data(fh: TextIO) -> list[list[int]]:
    return [list(map(int, line.split(' '))) for line in fh.read().splitlines()]


def compute(fn: Callable[..., int], fh: TextIO) -> str:
    return str(
        reduce(
            lambda x, y: list(map(add, map(fn, pairwise(x)), y)),
            reversed(parse_data(fh)),
        )[0]
    )


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p067_triangle.txt')
    result = compute(max, fh)
    fh.close()
    return result

File: clojure/src/project_euler/sol/p0067.clj
(ns project-euler.sol.p0067
  (:require
   [clojure.string :as str]
   [project-euler.lib.util :as util]))

(defn- parse-data
  [data]
  (->> (map #(str/split % #" ") data)
       (map #(map (fn [s] (parse-long s)) %))))

(defn solve
  ([]
   (solve "0067_triangle.txt"))
  ([fname]
   (->> (parse-data (util/read-data fname))
        (reverse)
        (reduce #(map + (map max %1 (rest %1)) %2))
        (first))))
